### ISSUE TITLE
broadcom_sta still fails to build (again, again) on kernel 5.18

### ISSUE THREAD

**Issue by NielsGx** [NixOS/nixpkgs#177798]
### Describe the bug
Building `broadcom_sta` fails when rebuilding system configuration.

### Steps To Reproduce
`boot.extraModulePackages = [ config.boot.kernelPackages.broadcom_sta ];`

### Expected behavior
System configuration should build successfully with `broadcom_sta` on kernel 5.18

### Screenshots
[err.txt](https://github.com/NixOS/nixpkgs/files/8912819/err.txt)
```
[...]
error: builder for '/nix/store/i2nsm53ln23dllrmxgf6cn3zrpan0h6y-broadcom-sta-6.30.223.271-5.18.3.drv' failed with exit code 2
error: 1 dependencies of derivation '/nix/store/qms8zvindb6d6i5ssavjw4zwswynl7sj-kernel-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/q2wll0bn3xys3z807d1c5dzbhkh559ss-linux-5.18.3-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/0kzp40rnjf1lrixm5jqz9y7szfnwgx0s-nixos-system-Niels-NixOS-22.05.915.9ff91ce2e4c.drv' failed to build
```

### Additional context
This appears to be a repetition of #177480 and #166035.
#177243 doesn't seem to be enough ?

I'm still on 21.11, can upgrade cleanly with its channel, but this happens when I try to rebuild after switching to 22.05 channel
(I did sudo nix-channel --update multiple times)

I'm still on kernel 5.17, but in my configuration.nix, I use the latest one, so I'm trying to upgrade to 5.18

I tried to use the arch patch (for 5.18, not for 5.17) talked about on #177480 but it still doesn't work, didn't checked if the error is the same though

### Notify maintainers
There are no current maintainers.

### Metadata
```console
- system: `"x86_64-linux"`
 - host os: `Linux 5.17.11, NixOS, 21.11 (Porcupine), 21.11.337851.9227bbe4315`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-22.05.915.9ff91ce2e4c"`
 - channels(niels): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by X9VoiD** [comment]
Sorry but I can't reproduce the issue on my system. On both 22.05 and unstable branch, using `pkgs.linuxPackages_latest` and `config.boot.kernelPackages.broadcom_sta`.

**Comment by icodeforyou-dot-net** [comment]
So in order to get my wifi up and running again, I ended up with `boot.kernelPackages = pkgs.linuxPackages;` instead of `pkgs.linuxPackages_latest` and using `boot.blacklistedKernelModules = [ "b43" "bcma" ];` in order to block these modules from loading. Because otherwise I had to unload those and then reload `wl` get wifi going again. I have no idea why I now all of a sudden had to do this blacklisting, but it is what it is...

**Comment by imxade** [comment]
> I ended up with `boot.kernelPackages = pkgs.linuxPackages;` instead of `pkgs.linuxPackages_latest` and using `boot.blacklistedKernelModules = [ "b43" "bcma" ];`

Thanks for the workaround, broadcom interface is visible now.

but unable to **scan** anything

```sh
$ lsmod | awk '/wl/'

wl                   6471680  0
cfg80211             1048576  1 wl

$ ip a

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: wlp6s0: <NO-CARRIER,BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state DORMANT group default qlen 1000
    link/ether 90:48:9a:f6:ae:8b brd ff:ff:ff:ff:ff:ff
    inet6 fe80::9248:9aff:fef6:ae8b/64 scope link 
       valid_lft forever preferred_lft forever
3: enp7s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
    link/ether b8:2a:72:c4:5c:8c brd ff:ff:ff:ff:ff:ff
4: enp0s20u1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN group default qlen 1000
    link/ether ca:47:dd:21:68:a2 brd ff:ff:ff:ff:ff:ff
    inet 192.168.62.25/24 brd 192.168.62.255 scope global dynamic noprefixroute enp0s20u1
       valid_lft 3261sec preferred_lft 2811sec
    inet6 fe80::c847:ddff:fe21:68a2/64 scope link 
       valid_lft forever preferred_lft forever

$ wpa_cli

Selected interface 'wlp6s0'

Interactive mode

<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
<3>CTRL-EVENT-SCAN-FAILED ret=-22 retry=1
```

**Comment by X9VoiD** [comment]
```nix
{
  boot.kernelModules = [ "wl" ];
  boot.extraModulePackages = [ config.boot.kernelPackages.broadcom_sta ];
}
```

This is the only thing I have for wifi and it builds correctly with `boot.kernelPackages = pkgs.linuxPackages_latest;`

I'm sorry but I still can't reproduce.

**Comment by imxade** [comment]
> ```nix
> {
>   boot.kernelModules = [ "wl" ];
>   boot.extraModulePackages = [ config.boot.kernelPackages.broadcom_sta ];
> }
> ```
> 
> This is the only thing I have for wifi and it builds correctly with `boot.kernelPackages = pkgs.linuxPackages_latest;` I'm sorry but I still can't reproduce the issue that you guys are experiencing.

indeed, that works

but still can't scan anything

**Comment by imxade** [comment]
UPDATE: 

**network manager** is working with **wpa supplicant** as backend.
Weirdly, On **replacing wpa supplicant** with **iwd** the wireless card isn't even detected.

**Comment by philiptaron** [comment]
I think I fixed it in #321134. Please take a look.
